backend/database-fs.js

```diff
@@ -36,38 +35,56 @@ async function testCaseSensitivity () {
   }
 }
 
-export async function initStorage (options: Object = {}): Promise<void> {
-  dataFolder = resolve(options.dirname)
-  await mkdir(dataFolder, { mode: 0o750, recursive: true })
-  await testCaseSensitivity()
-}
+export default class FsBackend {
+  dataFolder: string = ''
+  depth: number = 0
 
-// Useful in test hooks.
-export function clear (): Promise<void> {
-  return readdir(dataFolder)
-    .then(keys => Promise.all(keys.map(key => unlink(join(dataFolder, key)))))
-}
+  constructor (options: Object = {}) {
+    this.dataFolder = resolve(options.dirname)
+    this.depth = options.depth ?? 0
+    rebindMethods(this, ['init', 'clear', 'readData', 'writeData', 'deleteData'])
+  }
 
-// eslint-disable-next-line require-await
-export async function readData (key: string): Promise<Buffer | string | void> {
-  // Necessary here to thwart path traversal attacks.
-  checkKey(key)
-  return readFile(join(dataFolder, key))
-    .catch(err => undefined) // eslint-disable-line node/handle-callback-err
-}
+  // Maps a given key to a real path on the filesystem.
+  mapKey (key: string): string {
+    if (!this.depth) return join(this.dataFolder, key)
+    const keyChunks = key.match(/[A-Za-z]{1,3}/g) ?? []
+    return join(this.dataFolder, ...keyChunks.slice(0, this.depth), keyChunks.slice(this.depth).join(''))
+  }
 
-// eslint-disable-next-line require-await
-export async function writeData (key: string, value: Buffer | string): Promise<void> {
-  return writeFile(join(dataFolder, key), value)
-}
+  async init () {
+    await mkdir(this.dataFolder, { mode: 0o750, recursive: true })
+    await testCaseSensitivity(this)
+  }
 
-// eslint-disable-next-line require-await
-export async function deleteData (key: string): Promise<void> {
-  return unlink(join(dataFolder, key)).catch(e => {
-    // Ignore 'not found' errors
-    if (e?.code === 'ENOENT') {
-      return
-    }
-    throw e
-  })
+  async clear () {
+    await Promise.all(
+      await readdir(this.dataFolder)
+        .then(names => names.map(name => join(this.dataFolder, name)))
+        .then(paths => paths.map(path => rm(path, { recursive: true })))
+    )
+  }
+
+  async readData (key: string): Promise<Buffer | string | void> {
+    // Necessary here to thwart path traversal attacks.
+    checkKey(key)
+    return await readFile(this.mapKey(key))
+      .catch(err => undefined) // eslint-disable-line node/handle-callback-err
```

Comment:
Do you remember why we have this empty `.catch` to swallow errors?

We don't have one on `writeData`...

Are you sure it should be here? Does it have something to do with 404 handling?

Reply:
It's there to return `undefined` in case the file does not exist, i.e. if the given key is not found in the DB. It would be a bit more precise to do this only for `ENOENT` errors. Should we do this?

OTOH, `writeData` doesn't care whether a file is already there, as long as it can overwrite that file.
Other kinds of errors are possible, but we let them bubble up for now.

Reply:
> It would be a bit more precise to do this only for ENOENT errors. Should we do this?

I think it'd be more precise, although maybe `ENOENT` isn't the only error that should return undefined. 

Nitpick: If you'll swallow the error maybe add a comment explaining why, and maybe log something if the error isn't `ENOENT`.